pkg/session/state.go

```diff
@@ -209,85 +206,125 @@ func (spt *SessionState) RemoveFromRetrieval(storageProviderId peer.ID, retrieva
-	if protocol == multicodec.TransportGraphsyncFilecoinv1 {
-		gsmda := mda.(*metadata.GraphsyncFilecoinV1)
-		gsmdb := mdb.(*metadata.GraphsyncFilecoinV1)
-		// prioritize verified deals over not verified deals
-		if gsmda.VerifiedDeal != gsmdb.VerifiedDeal {
-			return gsmda.VerifiedDeal
-		}
+func (spt *SessionState) ChooseNextProvider(peers []peer.ID, mda []metadata.Protocol) int {
+	spt.lk.Lock()
+	defer spt.lk.Unlock()
 
-		// prioritize fast retrievel over not fast retrieval
-		if gsmda.FastRetrieval != gsmdb.FastRetrieval {
-			return gsmda.FastRetrieval
+	// score all peers, a float >=0 each, higher is better
+	scores := make([]float64, len(peers))
+	var tot float64
+	ind := make([]int, len(peers)) // index into peers that we can sort without changing peers
+	for ii, p := range peers {
+		ind[ii] = ii
+		gsmd, _ := mda[ii].(*metadata.GraphsyncFilecoinV1)
+		scores[ii] = spt.scoreProvider(p, gsmd)
+		tot += scores[ii]
+	}
+	// sort so that the non-random selection choose the first (best)
+	sort.Slice(ind, func(i, j int) bool {
+		return scores[ind[i]] > scores[ind[j]]
+	})
+
+	// choose a random peer, weighted by score
+	r := tot * spt.config.roll()
+	for _, pi := range ind {
+		s := scores[pi]
+		if r <= s {
+			return pi
 		}
+		r -= s
 	}
+	panic("unreachable")
+}
 
-	spt.lk.Lock()
-	defer spt.lk.Unlock()
-	sa := spt.spm[a]
-	sb := spt.spm[b]
+// scoreProvider returns a score for a given provider, higher is better.
+// This method assumes the caller holds lock.
+func (spt *SessionState) scoreProvider(id peer.ID, md *metadata.GraphsyncFilecoinV1) float64 {
+	var score float64
+	// var v, f bool
+
+	// graphsync metadata weighting
+	if md != nil {
+		if md.VerifiedDeal {
+			score += spt.config.GraphsyncVerifiedDealWeight
+			// v = true
+		}
+		if md.FastRetrieval {
+			score += spt.config.GraphsyncFastRetrievalWeight
+			// f = true
+		}
+	}
 
-	return sa.connectTime < sb.connectTime
+	// collected metrics scoring
+	sp := spt.spm[id]
+
+	// Exponential decay of connect time `f(x) = exp(-λx)` where λ is our
+	// exponential decay constant that we use to normalise the decay curve by
+	// observed connect time over all providers; giving us a stronger signal
+	// for relatively fast providers, but a long-tail toward a zero value for
+	// slow providers.
+	// https://en.wikipedia.org/wiki/Exponential_decay
+	// If we have no connect time data, use the average, if we have no average
+	// use 1.
+	λ := 1 / float64(spt.overallConnectTimeMs.getValue(1))
+	score += spt.config.ConnectTimeWeight * math.Exp(-λ*float64(sp.connectTimeMs.getValue(spt.overallConnectTimeMs.getValue(1))))
+
+	// if we have no success data, treat it as fully successful
+	score += spt.config.SuccessWeight * sp.success.getValue(1)
+
+	// DEBUG: fmt.Printf("%s v=%v, f=%v, oconn=%d, conn=%d, success=%f, score=%f\n", string(id), v, f, spt.overallConnectTimeMs.value, sp.connectTimeMs.value, sp.success.value, score)
```

Comment:
seems like this should be a log.Debug call rather than a commented out fmt.Printf